### PR TITLE
Added Slice documentation and new Slices examples.

### PR DESCRIPTION
**Pull request by PierreSachot**
Signed-off-by: Pierre Sachot <sachot.pierre@laposte.net>

### CODE REVIEW

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 1]
```diff
@@ -41,6 +43,90 @@
 	public void before() {
 		Utils.suppressSLF4JError();
 	}
+	
+	/**
+	 * Slice using basic 1D int array. The goal is to take the data from index 1
+	 * @throws Exception
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Remove @throws line in examples

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 2]
```diff
@@ -41,6 +43,90 @@
 	public void before() {
 		Utils.suppressSLF4JError();
 	}
+	
+	/**
+	 * Slice using basic 1D int array. The goal is to take the data from index 1
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Add to doc what the expected output is, match the javadoc added in Slice

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 3]
```diff
@@ -41,6 +43,90 @@
...
+	 * @throws Exception 
+	 */
+	@Test
+	public void sliceFrom1DFromIndex1() throws DatasetException {
+		final Dataset onedData = DatasetFactory.createFromObject(new int[]{1,2,3});
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Remove unneeded "final"s

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 4]
```diff
@@ -41,6 +43,90 @@
...
+	
+	
+	/**
+	 * Slice using basic 1D int array. The goal is to get the middle of the array
+	 * @throws Exception 
+	 */
+	@Test
+	public void sliceFrom1D() throws DatasetException {
+		final Dataset onedData = DatasetFactory.createFromObject(new int[]{10,9,8,7,6,5,4,3,2,1,0});
+		//we want to get the 3 values middle of the array, that's to say 6,5,4
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: reformat lines to be shorter (80-120 columns max)

Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: also comments should have space between `//` and text

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 5]
```diff
@@ -41,6 +43,90 @@
...
+		//This mean that you get the value at the 4th position to the value at the 6th position because the second one is exclusive, so you get at the index of end-1
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: This language is confusing between "position" and "index". IIRC position is normally 1-based, but index is 0-based.

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 6]
```diff
@@ -41,6 +43,90 @@
...
+	/**
+	 * Slice using basic 1D int array. The goal is to get from the middle of the array to the end
+	 * @throws Exception 
+	 */
+	@Test
+	public void sliceFrom1DGettingEnd() throws DatasetException {
+		final Dataset onedData = DatasetFactory.createFromObject(new int[]{10,9,8,7,6,5,4,3,2,1,0});
+		//we wanted to get the values from the middle, to the end-1, that's to say from 5 to 1
+		//to do that, we can do a simple slice like that :
+		Slice getter = new Slice(5, -1);
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Instead of -1 can you use `null`?

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 7]
```diff
@@ -41,6 +43,90 @@
...
+		//This mean that you get the value from the 5th position to the end, because -1 like in python represent the end here it will not get the 0 because the end is exclusive.
+		//and apply it to the Dataset, this will return the new Dataset :
+		Dataset sliceData = onedData.getSlice(getter);
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: inline `getter`

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 8]
```diff
@@ -41,6 +43,90 @@
...
+	 * @throws Exception 
+	 */
+	@Test
+	public void sliceFrom1D() throws DatasetException {
+		final Dataset onedData = DatasetFactory.createFromObject(new int[]{10,9,8,7,6,5,4,3,2,1,0});
+		//we want to get the 3 values middle of the array, that's to say 6,5,4
+		//This mean that you get the value at the 4th position to the value at the 6th position because the second one is exclusive, so you get at the index of end-1
+		//and apply it to the Dataset, this will return the new Dataset :
+		Dataset sliceData = onedData.getSlice(new Slice(4, 7));
+		//and show the informations :
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: remove this comment, it doesn't add much value as the next line is a `println`

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 9]
```diff
@@ -41,6 +43,90 @@
...
+	@Test
+	public void sliceFrom2D() throws DatasetException {
+		final Dataset twodData = DatasetFactory.createFromObject(new int[][]{	{1,2,3,4},
+																				{4,5,6,7},
+																				{7,8,9,10},
+																				{11,10,11,10}});
+		//we want to get the 4 values middle of the array, that's to say :	5,6
+																		// :	8,9
+		//to do that, we need to do a first slice for the first dimension, that's to say the rows :
+		Slice getterFirstD = new Slice(1,3);
```
Comment (jonahkichwacoders) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Can you update this example so that the two slices are different values please? e.g. return {6, 7}, {9, 10} instead.

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java (hunk 2 above)
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: change 'take' to slice

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 10]
```diff
@@ -43,6 +45,112 @@ public void before() {
 	}
 
 	/**
+	 * Slice using basic 1D int array. The goal is to take the data from index 1
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: change 'take' to 'create a Slice' so it matches the next method

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 11]
```diff
@@ -43,6 +45,112 @@ public void before() {
...
+
+	/**
+	 * Slice using basic 1D int array. The goal is to get the middle of the
+	 * array
+	 * This should display : [6, 5, 4]
+	 */
+	@Test
+	public void sliceFrom1D() {
+		Dataset onedData = DatasetFactory.createFromObject(new int[] { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 });
+		// we want to get the 3 values middle of the array, that's to say 6,5,4
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Simplify the comment to say

'We want to get the 4th to the 6th index. Start argument is inclusive so this will be 4. Stop argument is exclusive so this will be 7.'

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 12]
```diff
@@ -43,6 +45,112 @@ public void before() {
...
+		// the 6th index because the second one is exclusive, so you get at
+		// the index of end-1
+		// and apply it to the Dataset, this will return the new Dataset :
+		Dataset sliceData = onedData.getSlice(new Slice(4, 7));
+		System.out.println(sliceData.toString(true));
+	}
+
+	/**
+	 * Slice using basic 1D int array. The goal is to get from the middle of the
+	 * array to the end
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: this should say either say end-1 or the array should have a 0?

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 13]
```diff
@@ -43,6 +45,112 @@ public void before() {
...
+	 * This should display : [5, 4, 3, 2, 1]
+	 */
+	@Test
+	public void sliceFrom1DGettingEnd() {
+		Dataset onedData = DatasetFactory.createFromObject(new int[] { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 });
+		// we wanted to get the values from the middle, to the end-1, that's to
+		// say from 5 to 1
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: make this comment a bit more concise or maybe remove it?

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 14]
```diff
@@ -43,6 +45,112 @@ public void before() {
...
+	
+	/**
+	 * Slice using basic 1D int array. The goal is to get from the middle of the
+	 * array to the end
+	 * This should display : [5, 4, 3, 2, 1, 0]
+	 */
+	@Test
+	public void sliceFrom1DNullStop() {
+		Dataset onedData = DatasetFactory.createFromObject(new int[] { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0 });
+		// we wanted to get the values from the middle, to the end, that's to
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Make this comment more concise or maybe remove it?

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 15]
```diff
@@ -43,6 +45,112 @@ public void before() {
...
+		Dataset twodData = DatasetFactory
+				.createFromObject(new int[][] { { 1, 2, 3, 4 }, { 4, 5, 6, 7 }, { 7, 8, 9, 10 }, { 11, 10, 11, 10 } });
+		// we want to get the 4 values on the two middle lines and at the end of the line of the array, that's to say :
+		//	6,7
+		// 	9,10
+		// to do that, we need to do a first slice for the first dimension,
+		// that's to say the rows :
+		Slice getterFirstD = new Slice(1, 3);
+		// This means that you want to get the element from the index 1 to 2,
+		// because the second one is exclusive, so you get at the index of end-1
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Remove everything after the , (too verbose, let's not explain what exclusive means)

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 16]
```diff
@@ -43,6 +45,112 @@ public void before() {
...
+
+		// And you need a second slice to apply to the second dimension, here
+		// you want the 2nd and 3rd rows :
+		Slice getterSecondD = new Slice(2, 4);
+		// This means that you want to get the line at the index 2 to 3,
+		// because the second one is exclusive, so you get at the index of end-1
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: Remove everything after the , (too verbose, let's not explain what exclusive means)

File: org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java [hunk 17]
```diff
@@ -43,6 +45,112 @@ public void before() {
...
+
+		// and apply them to the Dataset, this will return the new Dataset :
```
Comment (tracymiranda) [inline org.eclipse.january.examples/src/org/eclipse/january/examples/dataset/SlicingExamples.java]: We can remove this, it just states the obvious

File: org.eclipse.january/src/org/eclipse/january/dataset/Slice.java [hunk 18]
```diff
@@ -76,14 +105,20 @@ private Slice(final Slice other) {
 	}
 
 	@Override
+	/**
+	 * Creates a deep copy of the Slice.
+	 * @return
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/Slice.java]: Remove newline/tabs between `@return` and doc (Try selecting the comment and pressing Ctrl-Shift-F to get correct formatting)

File: org.eclipse.january/src/org/eclipse/january/dataset/Slice.java [hunk 19]
```diff
@@ -242,12 +291,12 @@ public void setStep(int step) {
...
 	/**
-	 * Append string representation of slice
-	 * @param s
-	 * @param len
-	 * @param beg
-	 * @param end
-	 * @param del
+	 * Returns a String representation of the Slice comparable to the python representation.
+	 * @param s String builder
+	 * @param len Size of the wanted string
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/Slice.java]: len is the maximal length of the Slice, or -1 if not set

File: org.eclipse.january/src/org/eclipse/january/dataset/Slice.java [hunk 20]
```diff
@@ -356,12 +419,12 @@ public Slice flip() {
 	}
 
 	/**
-	 * Populate given start, stop, step arrays from given slice array 
-	 * @param shape
-	 * @param start
-	 * @param stop
-	 * @param step
-	 * @param slice
+	 * Populates given start, stop, step arrays from given slice array
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/Slice.java]: Document which of these parameters are input and which are output

File: org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java [hunk 21]
```diff
@@ -29,8 +29,8 @@
 	private boolean expanded;
 
 	/**
-	 * Construct ND slice for whole of shape
-	 * @param shape
+	 * Constructs a SliceND Object for whole of shape
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java]: Revert this change, the original was more correct.

File: org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java [hunk 22]
```diff
@@ -29,8 +29,8 @@
...
+	 * @param shape Array of shapes
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java]: should be `@param shape the shape of the dataset, see {@link ILazyDataset#getShape()}`

File: org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java [hunk 23]
```diff
@@ -45,19 +45,19 @@ public SliceND(final int[] shape) {
 	}
 
 	/**
-	 * Construct ND slice from an array of 1D slices
-	 * @param shape
-	 * @param slice
+	 * Constructs a SliceND Object from an array of 1D slices
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java]: As above

File: org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java [hunk 24]
```diff
@@ -45,19 +45,19 @@ public SliceND(final int[] shape) {
...
+	 * @param shape Array of shapes
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java]: As above

File: org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java [hunk 25]
```diff
@@ -45,19 +45,19 @@ public SliceND(final int[] shape) {
...
+	 * @param slice Slices array to add inside of the SliceND
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java]: should be `@param slice Slice for each dimension of ND slice` (actually improvements on wording welcome, but the idea is that each entry in the array applies to a dimension is important.

File: org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java [hunk 26]
```diff
@@ -45,19 +45,19 @@ public SliceND(final int[] shape) {
...
 	}
 
 	/**
-	 * Construct ND slice from an array of 1D slices
-	 * @param shape
-	 * @param maxShape can be null
-	 * @param slice
+	 * Constructs a SliceND Object from an array of 1D slices
+	 * @param shape Array of shapes
+	 * @param maxShape, may be {@code null}
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java]: If something can be null, it needs to say what happens in the null case.

File: org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java [hunk 27]
```diff
@@ -96,31 +96,25 @@ private void initMaxShape(int[] maxShape) {
...
 	 * 
-	 * @param shape
-	 * @param start
-	 *            can be null
-	 * @param stop
-	 *            can be null
-	 * @param step
-	 *            can be null
+	 * @param shape Array of shapes
+	 * @param start Array of starts points, may be {@code null}
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/SliceND.java]: Same thing here, if null what does null mean

File: org.eclipse.january/src/org/eclipse/january/dataset/Slice.java [hunk 28]
```diff
@@ -14,60 +14,89 @@
 
 import java.io.Serializable;
 
-/**
- * Class to represent a slice through a single dimension of a multi-dimensional dataset. A slice
- * comprises a starting position, a stopping position (not included) and a stepping size.
+/**	
+ * The {@code Slice} class represents the set of indices (start, stop, step), that are used to extract specifics subsets of {@link org.eclipse.january.dataset.Dataset}.<br><br>
+ * The start argument default to 0, stop argument default to the stop argument default to the end of the dimension that the slice is applied to, and the default argument for the step is 1.
```
Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/Slice.java]: The correct phrasing in English is:

The start argument defaults to 0
but
and the default argument for the step is 1

 :-)

Comment (jonahkichwacoders) [inline org.eclipse.january/src/org/eclipse/january/dataset/Slice.java]: >  stop argument default to the stop argument default to the end of the dimension

there are some duplicate words here
